packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,374 @@
+    if (
+      reference.identifier.range[0] < node.range[1] &&
+      reference.identifier.range[1] > node.range[0]
+    ) {
+      continue;
+    }
+
+    // Neither do references that aren't to the same type parameter,
+    // namely value-land (non-type) identifiers of the type parameter's type,
+    // and references to different type parameters or values.
+    if (
+      !reference.isTypeReference ||
+      reference.identifier.name !== node.name.name
+    ) {
+      continue;
+    }
+
+    // If the type parameter is being used as a type argument, then we
+    // know the type parameter is being reused and can't be reported.
+    if (
+      reference.identifier.parent.type === AST_NODE_TYPES.TSTypeReference &&
+      reference.identifier.parent.parent.type ===
+        AST_NODE_TYPES.TSTypeParameterInstantiation &&
+      reference.identifier.parent.parent.params.includes(
+        reference.identifier.parent,
+      )
+    ) {
+      return true;
+    }
+
+    total += 1;
+
+    if (total > 2) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Count uses of type parameters in inferred return types.
+ * We need to resolve and analyze the inferred return type of a function
+ * to see whether it contains additional references to the type parameters.
+ * For classes, we need to do this for all their methods.
+ */
+function countTypeParameterUsage(
+  checker: ts.TypeChecker,
+  node: NodeWithTypeParameters,
+): Map<ts.Identifier, number> {
+  const counts = new Map<ts.Identifier, number>();
+
+  if (ts.isClassLike(node)) {
+    for (const typeParameter of node.typeParameters) {
+      collectTypeParameterUsageCounts(checker, typeParameter, counts);
+    }
+    for (const member of node.members) {
+      collectTypeParameterUsageCounts(checker, member, counts);
+    }
+  } else {
+    collectTypeParameterUsageCounts(checker, node, counts);
+  }
+
+  return counts;
+}
+
+/**
+ * Populates {@link identifierCounts} by the number of times each type parameter
+ * appears in the given type by recursively descending through type references.
+ * This is essentially a limited subset of the scope manager, but for types.
+ */
+function collectTypeParameterUsageCounts(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  identifierCounts: Map<ts.Identifier, number>,
+): void {
+  const visitedSymbolLists = new Set<ts.Symbol[]>();
+  const type = checker.getTypeAtLocation(node);
+  const typeCounts = new Map<ts.Type, number>();
+  const visitedConstraints = new Set<ts.TypeNode>();
+  let functionLikeType = false;
+  let visitedDefault = false;
+
+  if (
+    ts.isCallSignatureDeclaration(node) ||
+    ts.isConstructorDeclaration(node)
+  ) {
+    functionLikeType = true;
+    visitSignature(checker.getSignatureFromDeclaration(node));
+  }
+
+  if (!functionLikeType) {
+    visitType(type, false);
+  }
+
+  function visitType(type: ts.Type | undefined, asRepeatedType: boolean): void {
```

Comment:
Continuing the discussion from https://github.com/typescript-eslint/typescript-eslint/pull/8173/#discussion_r1569362039

I spent some time stepping through this in the debugger trying to understand why `asRepeatedType` was necessary. The good news is that I get it now! The interesting news is that this rule works totally differently than how I expected. I think the way it's implemented is fine, but I think some clarifying comments and name changes could make it much clearer.

My main point of confusion was around what `collectTypeParameterUsageCounts` does, specifically what "recursively descending through type references" meant. I thought it meant that, when this function saw `Set<T>`, it would instantiate the type declaration for `Set` and see how many times `T` appeared in that. I don't think it does that, though! If it hits `Fn<Expr of T>`, then it will descend into `Expr of T` but not `Fn`. So `asRepeatedType` is a shortcut to make sure that `Fn<T>` always counts as a repeated use of `T`.

This isn't really correct since you could write code like this:

```ts
type Identity<T> = T;
declare function takeIdentity<T>(x: Identity<T>): void;
```

This is not a good use of generics, but the lint rule will allow it because of `asRepeatedType`.

I think this is an OK concession to expediency. But I can imagine real world examples of this getting thrown off by general helper types, e.g.:

```ts
type NonNullish<T> = T & {};
declare function foo<T>(x: NonNullish<T>): void;
```

So, a few comments / suggestions:

1. How hard would it be to recurse into the type aliases, not just the their type arguments? If this is impossible, then that's OK, but it would be good to at least note that (and add my two examples as tests!).
2. Clarify what is "recursively descended" into in the comment. Or maybe don't mention recursion at all, it's just walking the type equivalent of a syntax tree.
3. I think using the term "count" in this function name and its parameter (`identifierCounts`) is misleading since it doesn't really return a count. It _kind of_ returns a lower bound, but not really because of the `Identify` / `NonNullish` examples above. I'm not sure what a better name would be, maybe just something generic like `typeParameterWorker`?
4. For `asRepeatedType`, maybe the name could be `assumeMultipleUses`? Or maybe `asRepeatedType` is fine, just add a comment where you set it to `true`, e.g.:

```ts
// We don't know how many times the type argument is used in this type alias.
// Assume it's at least twice.
visitType(typeArgument, true);
```

Reply:
> 1. How hard would it be to recurse into the type aliases

I think it would be ... potentially doable, but very tricky. Users sometimes intentionally write types like `type PrintType<T> = T & {}` _(or something like that, I always forget the latest variants...)_. And then messaging to them that _"hey this type is indeed used, but not in a way we approve"_ is ... tough. 😬 

> 2. Clarify what is "recursively descended" into in the comment. Or maybe don't mention recursion at all, it's just walking the type equivalent of a syntax tree.

👍 agreed, that's a helpful suggestion. Switched to _"by checking its uses through its type references"_. If you have a better phrase I'd be interested in that.

> 3. `typeParameterWorker`

Yeah, this is tough... I'm not a fan of generic names (_"what does this mean?!"_) but am having a hard time finding one that's better suited here. Trying out, maybe, `foundIdentifierUsages`?

> 4. `assumeMultipleUses`

+1, I like that a lot more.